### PR TITLE
feat(vex): emit OpenVEX and CycloneDX VEX from scan results (#113)

### PR DESCRIPTION
**Pull request by lab700xdev**
Closes Lab700xOrg/aisbom-ops#113.

`aisbom scan --vex` writes VEX documents alongside the CycloneDX SBOM, stating per finding whether each scanned artifact is actually affected. FDA §524B has required VEX beside the SBOM since March 2026.

```bash
aisbom scan . --vex --output sbom.json
# -> sbom.json, sbom.openvex.json, sbom.vex.cdx.json
```

`--vex-format openvex|cyclonedx|both` (default `both`), `--vex-baseline <old-sbom.json>` for `fixed`.

## Design decisions

**Statements are keyed on AIsbom finding classes, not CVEs.** What AIsbom detects lives inside a model file — a pickle stream importing `os.system`, a Keras Lambda layer, a Jinja chat template with a sandbox escape. None of that has a CVE, because the file *is* the payload rather than a component with a patchable defect. Seven coarse, class-level identifiers, documented in [`docs/vex-finding-classes.md`](docs/vex-finding-classes.md), resolving under `https://aisbom.io/vex/`, never presented as CVE identifiers.

**`aliases` is deliberately empty.** The CVEs the bypass corpus knows — CVE-2025-1889/1944/1945, CVE-2025-10155/10156/10157, CVE-2025-1716 — are vulnerabilities in *picklescan*, another scanner failing to detect something. One entry literally reads "Fixed in picklescan 0.0.23." Aliasing an AIsbom finding to those would assert a relationship that does not exist, in a document a regulator may read. A test enforces it.

**Negatives are scoped to what was actually checked.** A class is asserted only for formats it can arise in, and only when the scan that would have found it ran to completion. Keras never receives a pickle-RCE claim at all: `_inspect_keras` reads the config with `scan_keras_config`, never `scan_pickle_stream`, so clearing one would assert a check that never ran.

**`fixed` requires `--vex-baseline`.** A single scan observes one point in time. The baseline is read from the SBOM's structured `aisbom:*` properties (#53/#54), falling back to the description string for older baselines — reading nothing there would report a clean baseline and turn every pre-existing finding into a spurious `fixed`.

**Justification is restricted to `vulnerable_code_not_present`.** OpenVEX offers five and CycloneDX nine, and they do not line up; mapping more would mean picking an inexact analogue for at least one. It is also the only justification a static read can support.

**CycloneDX VEX splits one entry per (class, status).** The library scopes one `analysis` block per vulnerability, but a repo routinely holds a malicious `.pt` and a clean `.safetensors` at once. Splitting keeps each component's status truthful.

**Compatibility policy, enforced.** A class is never deleted and an identifier never reused; a renamed class is aliased from its successor; `PUBLISHED_FINDING_CLASS_IDS` is append-only and a test fails if any published identifier stops resolving.

## Two defects found by leaving the fixtures

1. **Keras and ONNX over-claims.** A live `hf://google-bert/bert-base-uncased` scan showed `tf_model.h5` reported `not_affected` for Lambda execution when the config had hit the read budget — over a range request that is the *common* case for a `.h5`, not an edge case. Same for a truncated ONNX graph. Both now report `under_investigation`; regression test drives the real scanner.
2. **`KeyError` in the CycloneDX emitter.** Exercising the (currently empty) alias path revealed `generate_cyclonedx_vex` re-looking-up each class in the module-level registry instead of using the class the statement carries. It would have crashed on any statement AIsbom did not define — including the OSV-sourced CVE statements this module exists to accommodate. Fixed, with a test that serializes a `CVE-*` statement through both flavors.

## Verification

- `poetry run pytest --cov=aisbom --cov-fail-under=85` — **984 passed**, coverage **92.44%**, `aisbom/vex.py` at **100%**. 69 new tests.
- `aisbom bypass-scorecard --check` — gate passed, no case below the floor.
- **Live probe** `hf://google-bert/bert-base-uncased`: OpenVEX 0.2.0 strict validation PASS (official bundled schema, `additionalProperties: false`), CycloneDX 1.7 strict validation PASS, zero dangling product refs, all 4 model components covered.
- **Local probe** over `generate-test-artifacts`: `mock_malware.pt` `affected` with an action statement; SafeTensors and GGUF correctly `not_affected` with justification; real SHA-256 hashes emitted, sentinels never.
- **No regression:** insertion-only diff (0 deletions in every file). `aisbom scan .` against `origin/main` on the same tree — SBOM byte-identical (3095/3095), components identical, `bom-ref` set identical, terminal output identical, exit code identical.
- No dependency changes; `pyproject.toml` and `poetry.lock` untouched. +1.12 ms marginal import cost.
- CI gains smoke test #11 exercising both flavors and the baseline path.

## Follow-ups (not blockers)

- OSV/CVE mapping for `requirements.txt` pins — the CVE-keyed statements that join the same document. This is what actually closes the FDA §524B expectation of CVE-keyed VEX; the finding classes are additive to a submission, never a substitute.
- `https://aisbom.io/vex/*` pages the `@id` IRIs point at do not exist yet — fits the #115 compliance hub.
- `details["truncated"]` for ONNX is computed from the buffered read and stays `True` even when the streamed re-walk covered the file, so large *local* ONNX files report `under_investigation` conservatively. Making that field reflect the re-walk is a change to #105's output and deserves its own slice.
- GGUF `metadata_truncated` is not exposed as an `aisbom:*` property, so the baseline reader cannot see it. Harmless today (`fixed` only transitions from `affected`), but it is a gap in the structured surface.

### CODE REVIEW

File: aisbom/vex.py [hunk 1]
```diff
@@ -0,0 +1,933 @@
...
+def signals_from_artifact(art: Dict[str, Any]) -> Dict[str, Any]:
+    """Normalise a scanner artifact dict into the shared signal shape."""
+    sig = _empty_signals()
+    sig["format"] = _FRAMEWORK_TO_FORMAT.get(art.get("framework"))
+    details = art.get("details") or {}
+    fmt = sig["format"]
+
+    if fmt in _PICKLE_BEARING_FORMATS:
+        sig["pickle_opcodes"] = list(details.get("threats") or [])
+        sig["pickle_scan_incomplete"] = bool(details.get("scan_incomplete"))
```
[the remark this answers is not in the document]
Comment (lab700xdev) [reply aisbom/vex.py]: Confirmed and fixed in 17d9c7a — thank you, this was real.

Both branches you name set a "could not inspect" risk *label* without the structured marker `_classify` reads:

- `elif unreadable:` → `MEDIUM (Unscanned Container: {lz4,zstd})`
- `elif content is None and pickle_files:` → `MEDIUM (Unreadable Pickle Member)`

So every pickle class came out `not_affected` with justification `vulnerable_code_not_present` for a payload that was never opened.

Fixed at the source rather than in `vex.py`: both branches now set `details["scan_incomplete"]`, which is exactly what that field already means. Confining it to the VEX layer would have left the `--vex-baseline` reader still blind, since that reads the SBOM's `aisbom:*` properties rather than scanner internals.

Inspection errors, which you also flagged, had the same hole — a raised inspector leaves a sparse detail dict indistinguishable from a clean scan. `error` (artifact-level in most inspectors, under `details` in the GGUF one) now feeds the per-format incompleteness flags.

Regression test drives the real scanner: `test_unreadable_codec_is_not_cleared_in_vex_either` in `tests/test_pickle_variants.py`, parameterized over lz4 and zstd.

Measured SBOM impact: the component gains exactly one property, `aisbom:pickle:scan_incomplete=true`. Name, bom-ref, description, hashes and the risk label are unchanged, so exit codes are unaffected.

Worth noting this was the third instance of the same over-claim in this slice — Keras and ONNX were the first two, both found by scanning a live target.

File: aisbom/vex.py [hunk 2]
```diff
@@ -0,0 +1,933 @@
...
+    single scan observes one point in time.
+    """
+    statements: List[VexStatement] = []
+    for index, art in enumerate(artifacts):
+        sig = signals_from_artifact(art)
+        if sig["format"] is None:
+            continue
+        ref = bom_ref_for(index, art)
+        digest = _sha256_or_none(art.get("hash"))
+        prior = (baseline or {}).get(ref, {})
```
[the remark this answers is not in the document]
Comment (lab700xdev) [reply aisbom/vex.py]: Confirmed and fixed in 17d9c7a — and the second half of this finding is the one that mattered most.

`bom_ref_for` embeds the enumeration index (`artifact-<n>-<name>`). That is stable *within* a scan, which is all #111 needed it for, and not across two. As you say, adding a file earlier in the tree shifted every later index and turned a genuine `fixed` into `not_affected`.

The duplicate-basename case is worse than a miss, and it is reachable: artifact names are basenames, so `a/model.pt` and `b/model.pt` produce components differing only by index. If their order swaps, the affected baseline entry lands on the other, always-clean artifact and the document asserts a remediation that never happened — in the artifact whose entire purpose is evidencing remediation.

`baseline_findings` now returns a `BaselineIndex` with two identities, neither of them positional:

- `by_hash` — SHA-256, unforgeable but only proves byte-identity. Since remediation usually *changes* the file, this mainly disambiguates same-named artifacts rather than finding fixes.
- `by_name` — populated only for basenames that occur exactly once in the baseline; duplicates are dropped rather than guessed at.

The caller additionally requires the name to be unique in the *current* scan. Where identity cannot be established, no prior is claimed at all: a missed `fixed` costs the document a nice-to-have, while a mis-attributed one is a false statement.

Three tests cover it, including the exact scenario you describe:

- `test_a_fix_is_still_recognised_when_an_earlier_file_shifts_the_index`
- `test_duplicate_basenames_never_fabricate_a_fix`
- `test_a_matching_content_hash_resolves_an_ambiguous_name`

Noting one residual limitation for the record: the scanner stores basenames and no scan-root-relative path, so two same-named artifacts with no hashes (the remote-scan shape, where `remote_unhashed` is the sentinel) remain genuinely unidentifiable across scans. That case now yields no prior rather than a guess.
